README.md
# ipfs-provider

[![](https://img.shields.io/github/release/ipfs-shipyard/ipfs-provider.svg)](https://github.com/ipfs-shipyard/ipfs-provider/releases/latest)
[![](https://img.shields.io/badge/made%20by-Protocol%20Labs-blue.svg?style=flat-square)](https://protocol.ai)
[![](https://img.shields.io/badge/project-IPFS-blue.svg?style=flat-square)](http://ipfs.io/)
[![Build Status](https://flat.badgen.net/travis/ipfs-shipyard/ipfs-provider)](https://travis-ci.com/ipfs-shipyard/ipfs-provider)
[![Dependency Status](https://david-dm.org/ipfs-shipyard/ipfs-provider.svg?style=flat-square)](https://david-dm.org/ipfs-shipyard/ipfs-provider)

> Returns IPFS API object by trying multiple [providers](#providers) in a custom fallback order.
>
> This is a general-purpose replacement for [ipfs-redux-bundle](https://github.com/ipfs-shipyard/ipfs-redux-bundle).

- [Install](#install)
- [Usage](#usage)
- [Examples](#examples)
- [Providers](#providers)
  - [`httpClient`](#httpclient) (remote/local HTTP API)
  - [`jsIpfs`](#jsipfs) (embedded js-ipfs node)
<!-- TODO: improve this API  - [create our own!](#creating-a-provider) -->
- [Test](#test)


## Install

### via NPM

```console
$ npm install ipfs-provider
```

### via prebuilt browser bundles

```html
<!-- remember to include js-ipfs (core) and/or js-ipfs-http-client, if they are used -->
<script src="https://cdn.jsdelivr.net/npm/ipfs-core/dist/index.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/ipfs-http-client/dist/index.min.js"></script>

<!-- prebuilt and minified bundle -->
<script src="https://cdn.jsdelivr.net/npm/ipfs-provider/dist/index.min.js"></script>

<script>
  const { getIpfs, providers } = window.IpfsProvider
  const { ipfs, provider, apiAddress } = await getIpfs({
    loadHttpClientModule: () => window.IpfsHttpClient,
    loadJsIpfsModule: () => window.IpfsCore,
    providers: [ /* see Usage below */ ]
    })
</script>
```

**Note:** when using prebuilt bundles in production use explicit versions and SRI hashes.
Details [here](https://github.com/ipfs/js-ipfs/tree/master/packages/ipfs-http-client#in-a-web-browser).
<!-- TODO: add prebuild+versions+sri example to ./examples -->

## Usage

```js
const { getIpfs, providers } = require('ipfs-provider')
const { httpClient, jsIpfs } = providers

const { ipfs, provider, apiAddress } = await getIpfs({
  // when httpClient provider is used multiple times
  // define its constructor once, at the top level
  loadHttpClientModule: () => require('ipfs-http-client'),

  // note this is an array, providers are tried in order:
  providers: [

    // try various HTTP endpoints (best-effort),
    httpClient({
      // (1) try multiaddr of a local node
      apiAddress: '/ip4/127.0.0.1/tcp/5001'
    }),
    httpClient(), // (2) try "/api/v0/" on the same Origin as the page
    httpClient({
      // (3) try arbitrary API from URL string
      apiAddress: 'https://some.example.com:8080'
    }),
    httpClient({
      // (4) try API defined by a custom http client config
      apiAddress: {
        host: 'apis.example.com',
        port: '443',
        protocol: 'https',
        apiPath: 'custom/path/to/api/v0',
        headers: {
          authorization: 'Basic dXNlcjpwYXNz'
        }
      }
    }),
    // (5) final fallback to spawning embedded js-ipfs running in-page
    jsIpfs({
      // js-ipfs package is used only once, as a last resort
      loadJsIpfsModule: () => require('ipfs-core'),
      options: { } // pass config: https://github.com/ipfs/js-ipfs/blob/master/packages/ipfs/docs/MODULE.md#ipfscreateoptions
    })
  ]
})

for await (const file of ipfs.add("text")) {
  if (file && file.cid) {
    console.log(`successfully stored at ${file.cid}`)
  } else {
    console.error('unable to ipfs.add', file)
  }
}
```

- `ipfs` – returned instance of IPFS API (see [SPEC](https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/README.md))
- `provider` – a string with a name of the first successful provider.
   - built-in names match constants from `providers`: `httpClient`, `jsIpfs`, `windowIpfs` and `webExt`.
- `apiAddress` – returned only when `httpClient` provider is used, provides information which HTTP endpoint succeded


## Examples

See [`examples/`](./examples) for sample code and demonstration of advanced fallback strategies.

## Providers

You can customize the order of the providers by passing a different array order to the `providers` array.


For example, if you want to try `httpClient` and then `jsIpfs`, you should run it like this:

```js
const { getIpfs, providers } = require('ipfs-provider')
const { httpClient, jsIpfs } = providers

const { ipfs, provider } = await getIpfs({
  providers: [
    httpClient(),
    jsIpfs()
  ]
})
```

#### Customizing connection test

```js
const { ipfs, provider } = await getIpfs({
  providers: [ /* array of providers to try in order */ ],
  connectionTest: () => { /* boolean function to test the connection to IPFS, default one tries to ipfs.get the CID of an empty directory */ },
})
```

### `httpClient`

Tries to connect to HTTP API via [`js-ipfs-http-client`](https://github.com/ipfs/js-ipfs/tree/master/packages/ipfs-http-client):

```js
const { ipfs, provider } = await getIpfs({
  providers: [
    httpClient({
      loadHttpClientModule: () => require('ipfs-http-client'),
      apiAddress: 'https://api.example.com:8080/'
    })
  ]
})
```

This provider will attempt to establish connection with (in order):
1. `apiAddress` (if provided)
2. `/api/` at the current Origin
3. the default local API (`/ip4/127.0.0.1/tcp/5001`)

It supports lazy-loading and small bundle sizes. The client library is initialized using constructor (in order):
1. one returned by `loadHttpClientModule` async function (if provided)
2. one exposed at `window.IpfsHttpClient` (if present)

Value passed in `apiAddress` can be:
- a multiaddr (string like `/ip4/127.0.0.1/tcp/5001` or an [object](https://github.com/multiformats/js-multiaddr/))
- a String with an URL (`https://api.example.com:8080/`)
- a configuration object supported by the [constructor](https://github.com/ipfs/js-ipfs/tree/master/packages/ipfs-http-client#importing-the-module-and-usage)
  (`{ host: '1.1.1.1', port: '80', apiPath: '/ipfs/api/v0' }`)


To try multiple endpoints, simply use this provider multiple times.
See [`examples/browser-browserify/src/index.js`](./examples/browser-browserify/src/index.js) for real world example.

### `jsIpfs`

Spawns embedded [`js-ipfs`](https://github.com/ipfs/js-ipfs/tree/master/packages/ipfs) (a full IPFS node in JavaScript)
in the context of the current page using customizable constructor:

```js
const { ipfs, provider } = await getIpfs({
  providers: [
    jsIpfs({
      loadJsIpfsModule: () => require('ipfs-core'),
      options: { /* advanced config */ }
    })
  ]
})
```

- `loadJsIpfsModule` should be a function that returns a promise that resolves to a js-ipfs constructor
   <!-- TODO confirm below is true, if it is, add example to examples/ and link to it
   This works well with [dynamic `import()`](https://developers.google.com/web/updates/2017/11/dynamic-import), so you can lazily load js-ipfs when it is needed.
   -->
- `options` should be an object which specifies [advanced configurations](https://github.com/ipfs/js-ipfs#ipfs-constructor) to the node.
- **TIP:** when used in a browser context, use `ipfs-core` for a smaller browser bundle

### `windowIpfs`

[`window.ipfs`](https://github.com/ipfs-shipyard/ipfs-companion/blob/master/docs/window.ipfs.md) was an experiment created by [ipfs-companion](https://github.com/ipfs/ipfs-companion) browser extension.
It supported passing an optional list of permissions to [display a single ACL prompt](https://github.com/ipfs-shipyard/ipfs-companion/blob/master/docs/window.ipfs.md#do-i-need-to-confirm-every-api-call) the first time it is used:

```js
const { ipfs, provider } = await getIpfs({
  providers: [
    windowIpfs({
      // example
      permissions: { commands: ['add','cat','id', 'version'] }
    })
  ]
})
```

### `webExt`

`webExt` looks for an instance in the [background page of a WebExtension](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/extension/getBackgroundPage)
(useful only in browser extensions, not regular pages, disabled by default)

```js
const { ipfs, provider } = await getIpfs({
  providers: [
    webExt()
  ]
})
```

<!-- TODO: improve this API

#### Creating a provider

If built-in providers do not solve your use case, create your own!

A provider is just a function that returns an async function that returns
`{ ipfs, provider }`, but we provide `makeProvider` which enables `connectionTest`:

```js
const { getIpfs, makeProvider } = require('ipfs-provider')

const customProvider = makeProvider(async ({ connectionTest }) => {
  try {
    const ipfs = // create IPFS API instance somehow
    await connectionTest(ipfs)
    return { ipfs, provider: 'customProvider' }
  } catch (_) {
    return null
  }
})

const { ipfs, provider } = await getIpfs({
  providers: [
    customProvider()
  ]
})
```

-->

## Test

```console
$ npm test
```

## Lint

Perform [`standard`](https://standardjs.com/) linting on the code:

```console
$ npm run lint
```

## Contribute

Feel free to dive in! [Open an issue](https://github.com/ipfs-shipyard/ipfs-provider/issues/new) or submit PRs.

To contribute to IPFS in general, see the [contributing guide](https://github.com/ipfs/community/blob/master/CONTRIBUTING.md).

[![](https://cdn.rawgit.com/jbenet/contribute-ipfs-gif/master/img/contribute.gif)](https://github.com/ipfs/community/blob/master/CONTRIBUTING.md)

## License

[MIT](LICENSE) © Protocol Labs


CHANGELOG.md
# [2.1.0](https://github.com/ipfs-shipyard/ipfs-provider/compare/v2.0.1...v2.1.0) (2021-09-01)


### Features

* support 'create' constructor from ipfs-http-client ([#38](https://github.com/ipfs-shipyard/ipfs-provider/issues/38)) ([df7b9b3](https://github.com/ipfs-shipyard/ipfs-provider/commit/df7b9b3cc7423a4ac894dbec813eb3c0a65fef2f))



## [2.0.1](https://github.com/ipfs-shipyard/ipfs-provider/compare/v2.0.0...v2.0.1) (2021-04-02)



# [2.0.0](https://github.com/ipfs-shipyard/ipfs-provider/compare/v1.1.0...v2.0.0) (2021-04-02)


### Features

* add options to http client ([#30](https://github.com/ipfs-shipyard/ipfs-provider/issues/30)) ([4115818](https://github.com/ipfs-shipyard/ipfs-provider/commit/41158187fcc3153a0824b2cc871da7c769879415))



<a name="1.1.0"></a>
# [1.1.0](https://github.com/ipfs-shipyard/ipfs-provider/compare/v1.0.0...v1.1.0) (2020-09-09)


### Features

* support custom config for ipfs-http-client ([#24](https://github.com/ipfs-shipyard/ipfs-provider/issues/24)) ([319ba34](https://github.com/ipfs-shipyard/ipfs-provider/commit/319ba34))



<a name="1.0.0"></a>
# [1.0.0](https://github.com/ipfs-shipyard/ipfs-provider/compare/v0.2.2...v1.0.0) (2020-04-22)


### Bug Fixes

* run jest before aegir release ([0e40e66](https://github.com/ipfs-shipyard/ipfs-provider/commit/0e40e66))


### Features

* better api ([#10](https://github.com/ipfs-shipyard/ipfs-provider/issues/10)) ([58098a8](https://github.com/ipfs-shipyard/ipfs-provider/commit/58098a8))
* lazy load js-ipfs-http-client and async iterator support ([#8](https://github.com/ipfs-shipyard/ipfs-provider/issues/8)) ([cac51fa](https://github.com/ipfs-shipyard/ipfs-provider/commit/cac51fa))
* prebuilt browser bundles ([bc1c63d](https://github.com/ipfs-shipyard/ipfs-provider/commit/bc1c63d))
* **BREAKING:** ipfs-http-client v44 ([#19](https://github.com/ipfs-shipyard/ipfs-provider/issues/19)) ([8d9d025](https://github.com/ipfs-shipyard/ipfs-provider/commit/8d9d025))


### BREAKING CHANGES

* this switches ipfs-provider to the new JS APIs based on
async iterators. More details in: https://blog.ipfs.io/2020-02-01-async-await-refactor
* **BREAKING:** ipfs-http-client v44 brings new JS API (with async iterators)



<a name="0.2.2"></a>
## [0.2.2](https://github.com/ipfs-shipyard/ipfs-provider/compare/v0.2.1...v0.2.2) (2019-09-02)


### Features

* support window.ipfs.enable() ([#5](https://github.com/ipfs-shipyard/ipfs-provider/issues/5)) ([b720386](https://github.com/ipfs-shipyard/ipfs-provider/commit/b720386)), closes [#1](https://github.com/ipfs-shipyard/ipfs-provider/issues/1)



<a name="0.2.1"></a>
## [0.2.1](https://github.com/ipfs-shipyard/ipfs-provider/compare/v0.2.0...v0.2.1) (2019-06-14)


### Bug Fixes

* do not fail outside the browser ([2714381](https://github.com/ipfs-shipyard/ipfs-provider/commit/2714381))



<a name="0.2.0"></a>
# [0.2.0](https://github.com/ipfs-shipyard/ipfs-provider/compare/v0.1.0...v0.2.0) (2019-06-07)


### Features

* add tests and output the provider ([eb80656](https://github.com/ipfs-shipyard/ipfs-provider/commit/eb80656))



<a name="0.1.0"></a>
# [0.1.0](https://github.com/ipfs-shipyard/ipfs-provider/compare/1c25ef9...v0.1.0) (2019-06-05)


### Bug Fixes

* add missing dependencies ([1c25ef9](https://github.com/ipfs-shipyard/ipfs-provider/commit/1c25ef9))
* early return of the first provider ([22dfb4e](https://github.com/ipfs-shipyard/ipfs-provider/commit/22dfb4e))
* validate apiAddress option ([a58d6ec](https://github.com/ipfs-shipyard/ipfs-provider/commit/a58d6ec))





examples/browser-esbuild/src/index.js
'use strict'

const { getIpfs, providers } = require('ipfs-provider')
const { httpClient, jsIpfs } = providers

document.addEventListener('DOMContentLoaded', async () => {
  console.log('IPFS starting..')
  const { ipfs, provider, apiAddress } = await getIpfs({
    // HTTP client library can be defined globally to keep code minimal
    // when httpClient provider is used multiple times
    loadHttpClientModule: () => require('ipfs-http-client'),
    // try window.ipfs (if present),
    // then http apis (if up),
    // and finally fallback to spawning embedded js-ipfs
    providers: [
      httpClient({
        // try multiaddr of a local node
        apiAddress: '/ip4/127.0.0.1/tcp/5001'
      }),
      httpClient(), // try "/api/v0/" on the same Origin as the page
      httpClient({
        // try URL of a remote node
        apiAddress: 'http://dev.local:8080'
      }),
      jsIpfs({
        // js-ipfs package is used only once, here
        loadJsIpfsModule: () => require('ipfs-core'), // note require instead of
        options: {
          init: { alogorithm: 'ed25519' }
        }
      })
    ]
  })

  console.log('IPFS API is provided by: ' + provider)
  if (provider === 'httpClient') {
    console.log('HTTP API address: ' + apiAddress)
  }

  async function store () {
    const toStore = document.getElementById('source').value
    const result = await ipfs.add(toStore)
    if (result && result.cid) {
      console.log('successfully stored', result)
      await display(result.cid.toString())
    } else {
      console.error('unable to ipfs.add', result)
    }
  }

  async function display (cid) {
    const chunks = []
    for await (const chunk of ipfs.cat(cid)) {
      chunks.push(chunk)
    }

    // merge all chunks into single byte buffer (does not matter here, as text
    // entered by user will be short enough to fit in a single chunk, but makes
    // this example useful for bigger files)
    const data = new Uint8Array(chunks.reduce((acc, curr) => [...acc, ...curr], []))

    document.getElementById('cid').innerText = cid
    document.getElementById('content').innerText = new TextDecoder().decode(data)
    document.getElementById('output').setAttribute('style', 'display: block')
  }

  document.getElementById('store').onclick = store
})
